jax/linear_util.py

```diff
@@ -208,15 +196,16 @@ def wrap_init(f, params={}):
   return WrappedFun(f, (), (), tuple(sorted(params.items())))
 
 
-def cache(call, max_size=4096):
-  @fastcache.clru_cache(maxsize=max_size)
-  def cached_fun_body(f, args):
-    return call(f, *args), f
-
-  def cached_fun(f, *args):
-    ans, f_prev = cached_fun_body(f, args)
-    if id(f_prev) != id(f):
-      f.populate_stores(f_prev)
+def cache(call):
+  caches = weakref.WeakKeyDictionary()
+  def memoized_fun(f, *args):
+    cache = caches.setdefault(f.f, {})
+    key = (f.transforms, f.params, args)
+    if key in cache:
+      ans, stores = cache[key]
```

Comment:
I'm confused why this works at all. Doesn't key (the tuple) lose its only strong reference at the end of this function? Wouldn't this mean that all functions can and will always be evicted?

Reply:
Ok, I read this more carefully. It's fine. But the difference between `caches` and `cache` is easy to miss on a quick skim.